File: Data Structures & Algorithms/invert-a-binary-tree/submission-2.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Solution {
    fun invertTree(root: TreeNode?): TreeNode? {
        fun helper(root: TreeNode?): TreeNode?{
            if(root == null)
                return root
            
            val temp = root.left  
            root.left = root.right
            root.right = temp

            helper(root.left)
            helper(root.right)

            return root
        }

        return helper(root)
    }
}

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def buildTree(self, preOrder: List[int], inOrder: List[int]) -> Optional[TreeNode]:
        inorder_map = defaultdict()
        for i,v in enumerate(inOrder):
            inorder_map[v] = i

        ind=0
        def build(low, high):
            nonlocal ind
            if low > high:
                return

            val = preOrder[ind]
            ind += 1
            
            root = TreeNode(val)
            root.left = build(low, inorder_map[val]-1)
            root.right = build(inorder_map[val]+1, high)
            
            return root
        
        return build(0, len(preOrder)-1)
                        
